### DISCUSSION TITLE
Different builddir for files in same source folder

### DISCUSSION THREAD

**Discussion by connectrajeev** [mesonbuild/meson#13633] [General]
Hi All, 

   I am new to meson buildtool and learning it these days. I have a specifc requirement where I want to generate two builddir let's say builddir1 & builddir2 from one source folder which contains all C files.  So some of the object files should be generated in builddir1 and some should be in builddir2 - I can define source_builddir1 and source_builddir2 seprately and finally I want to generate either a static library or a shared library or an executable (I can define that too in meson.build) from these object files and store it in the builddir1 folder. 
I don't think there is any in-built function in the meson which can perfrom like this so I have to write a custom function for this can someone let me know how can I achive this.
All source files are in source_dir 

source_dir = subdir/
builddir1 = subdir/dir1
builddir2  = subdir/dir2

source_builddir1 = files('a.c', 'b.c', 'c.c', )
source_builddir2 = files('d.c', 'e.c',)

 Thanks,
-Rajeev

**Comment by rilian-la-te** [comment]
https://gitlab.com/vala-panel-project/vala-panel-appmenu/-/tree/master/subprojects/appmenu-gtk-module/src?ref_type=heads - look here, exactly your goal.

I created 2 subdirs in source just with a meson.build files, and they are build different binaries from one source folder.

**Comment by eli-schwartz** [comment]
> So some of the object files should be generated in builddir1 and some should be in builddir2 - I can define source_builddir1 and source_builddir2 seprately and finally I want to generate either a static library or a shared library or an executable (I can define that too in meson.build) from these object files and store it in the builddir1 folder.

I am not entirely certain what you are trying to do here, but to shed a bit of light on how meson operates:

- there is only one project_build_root, that is where you configure the build in
  - you can configure multiple build profiles by specifying different builddirs to `meson setup`, for example if you want to build your project using both debug and release toolchains, or do a build with clang and a build with gcc, or build one native version of your project and one version that is compiled using mingw to target Windows
- if you want to put libraries and executables in `project_build_root() / 'dir1'`, you can do that by having dir1/meson.build which defines your `executable()`. It can use files from any directory as *source inputs*. Either by specifying them as `executable('myprog', '../source_a.c')` or by defining, in the top sourcedir's meson.build file, `all_sources = files('source_a.c')` and reusing that "files" object
